diffedit.md

```diff
@@ -0,0 +1,752 @@
+combined_embeddings = create_text_embeddings(prompts)
+```
+
+## Noising the original image
+
+To add noise to our original image to the correct level, we will utilise the scheduler. The scheduler typically manages noise throughout the image creation process for diffusion models, and it can also be used to manually add a specific level of noise to an image.
+
+First we need to set up the scheduler, and provide it with the configuration values that our diffusion model expects. We have used a DDIM scheduler, which is the type of scheduler that is recommended by the DiffEdit paper as it allows for more consistent results between the original image and the edited image.
+
+We'll define this setup as a function as we'll be using it again later when we generate the DiffEdit image. The argument `num_inference_steps` is the total number of steps that our model would take to fully denoise an image. With the version of Stable Diffusion and the scheduler we have chosen, 50 steps usually works well.
+
+
+```python
+scheduler_config = {
+  "beta_end": 0.012,
+  "beta_schedule": "scaled_linear",
+  "beta_start": 0.00085,
+  "num_train_timesteps": 1000,
+  "set_alpha_to_one": False,
+  "steps_offset": 1,
+  "trained_betas": None,
+  "clip_sample": False
+}
+```
+
+
+```python
+def init_scheduler(num_inference_steps):  
+    scheduler = DDIMScheduler(**scheduler_config)
+    scheduler.set_timesteps(num_inference_steps)
+    return scheduler
+```
+
+
+```python
+num_inference_steps = 50
+scheduler = init_scheduler(num_inference_steps)
+```
+
+Next, we will define a generic function which creates noise and adds it to the latents we pass in. It also takes in a start step for the scheduler which allows us to control the amount of noise added to the image. 
+
+
+```python
+def create_noise(scheduler, initial_latents, height=512, width=512, batch_size=1, start_step=10, seed=None):
+    if seed is not None:
+        torch.manual_seed(seed)
+    initial_noise = torch.randn((batch_size, unet.config.in_channels, height // 8, width // 8)).to("cuda").half()
+    
+    return scheduler.add_noise(
+        initial_latents,
+        initial_noise, 
+        timesteps=torch.tensor([scheduler.timesteps[start_step]])
+    )
+```
+
+We will pass in the latents from our encoded original image into our new function, and we want to add 50% noise as suggested in the DiffEdit paper. Our function doesn't take in an amount of noise specifically, but allows us to specify which step the scheduler should use to noise the image. Halfway through the total number of steps will add 50% noise.
+
+
+```python
+mask_noise_step = num_inference_steps//2
+noisy_latents = create_noise(scheduler, original_image_latents, start_step=mask_noise_step)
+```
+
+We can then take a look at our noised latents to get an idea of what it they look like:
+
+
+```python
+latents_to_grid_bw(noisy_latents.squeeze().detach().cpu(), 4)
+```
+
+
+    
+![png](diffedit_files/diffedit_43_0.png)
```

Comment:
All of these and the subsequent images do a great job of visualising your progress.